### PR TITLE
[FEAT] [EXPERIMENTAL] added support for next_by_subj to direct stream api

### PR DESCRIPTION
**Pull request by aricart**
[REFACTOR] moved direct API to its own JSM api class

### CODE REVIEW

File: nats-base-client/jsmdirect_api.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  get time(): Date {
+    return new Date(this.header.get(RepublishedHeaders.JsTimeStamp));
+  }
+
+  get stream(): string {
+    return this.header.get(RepublishedHeaders.JsStream);
+  }
+
+  get lastSequence(): number {
+    const v = this.header.get(RepublishedHeaders.JsLastSequence);
```
Comment (kozlovic) [inline nats-base-client/jsmdirect_api.ts:85]: I am not seeing that being set by the server. Where would that come from?

**Review by kozlovic — COMMENTED** [review]
See my latest comment in the ADR regarding what needs to be returned to the user and question about "NATS" direct get specific headers..

File: nats-base-client/jsmdirect_api.ts (hunk 1 above)
Comment (aricart) [reply nats-base-client/jsmdirect_api.ts]: very good catch - this was related to the RepublishedMessageHeaders which is not what this PR is about.

**Review by kozlovic — CHANGES_REQUESTED** [review]
A bit confused on what DirectMsgRequest is, since I don't see how this is used. Also looks like NextMsgRequest is missing a field? (but again, not sure how this is being used).

File: nats-base-client/types.ts [hunk 2]
```diff
@@ -469,6 +469,27 @@ export interface StreamAPI {
...
+ * sequence is returned.
+ * @field seq - the seq to start looking
+ * @field next_by_subject - the subject to look for
+ */
+export type NextMsgRequest = {
+  seq: number;
+  next_by_subj: string;
+};
+
+export type DirectMsgRequest =
```
[the remark this answers is not in the document]
Comment (aricart) [reply nats-base-client/types.ts:484]: The API returns a `StoredMsg` which is an interface (same as what you can `$JS.API.STREAM.MSG.GET.${stream}`. In JavaScript this is typed, as you cannot ack or do anything with these messages to a data type.

File: nats-base-client/types.ts [hunk 3]
```diff
@@ -469,6 +469,27 @@ export interface StreamAPI {
...
+/**
+ * Request the next stream message by sequence for the specified subject.
+ * Note that if the specified seq matches the specified subject, that
+ * sequence is returned.
+ * @field seq - the seq to start looking
+ * @field next_by_subject - the subject to look for
+ */
+export type NextMsgRequest = {
+  seq: number;
+  next_by_subj: string;
```
[the remark this answers is not in the document]
Comment (aricart) [reply nats-base-client/types.ts:481]: It does, the argument that you can provide for getMessage as a `DirectMsgRequest` which is an alias for one of  three types `LastForMsgRequest` is one of the possible values: 
```
export type DirectMsgRequest =
  | SeqMsgRequest
  | LastForMsgRequest
  | NextMsgRequest;
```

File: nats-base-client/types.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (aricart) [reply nats-base-client/types.ts:484]: (typescript is just looking for a type that has the fields defined by the type) it is not cast/it is just something that has that 'shape'.

[the remark this answers is not in the document]
Comment (kozlovic) [reply nats-base-client/types.ts:484]: We followed up on a call and it is clear to me now.

**Review by kozlovic — APPROVED** [review]
LGTM

**Pull request merged**